component/config.go

```diff
@@ -31,46 +33,131 @@ type ConfigValidator interface {
 // ValidateConfig validates a config, by doing this:
 //   - Call Validate on the config itself if the config implements ConfigValidator.
 func ValidateConfig(cfg Config) error {
-	return validate(reflect.ValueOf(cfg))
+	var err error
+	errs := validate(reflect.ValueOf(cfg))
+
+	for _, suberr := range errs {
+		if suberr.err != nil {
+			if suberr.path != "" {
+				err = errors.Join(err, fmt.Errorf("%s: %w", suberr.path, suberr.err))
+			} else {
+				err = errors.Join(err, suberr.err)
+			}
+		}
+	}
+
+	return err
+}
+
+type pathError struct {
+	err  error
+	path string
```

Comment:
I would store this as a slice here
```suggestion
	path []string
```

and then do the joining in the `Error` method, I think that will give you simpler code (you avoid all the ifs).

You can probably store in reverse order as well so that you just append to the list, and then reverse when printing

Reply:
I opted for a different name, but I agree we should just join at the end. This also cleans up the code a good amount, thanks for calling this out.

Reply:
> Edit: To clarify, I went with toError instead of Error so as not to conflict with the error interface, which specifies that Error returns a string.

Oops, I missed this and I left a suggestion below to change it to implement the `Error` interface :smile: